designs/2022-languages/README.md

```diff
@@ -0,0 +1,610 @@
+
+# ESLint Language Plugins
+
+## Summary
+
+This proposal provides a way for plugin authors to define parsers, rules, etc., for any arbitrary programming languages. This would allow ESLint to work on any type of programming language for which there is a plugin defined, not just for JavaScript and its variants.
+
+## Motivation
+
+Over the nine years ESLint has been in existence, it has grown into being not just the most popular JavaScript linter, but arguably the most popular linter for any programming language. Twitter is filled with people lamenting that they don't have an "ESLint for X", where X is some other file type they're working with. Numerous projects have popped up that mimic a lot of what ESLint does at it's core:
+
+* [Stylelint](https://stylelint.io/);
+* [Markdownlint](https://github.com/DavidAnson/markdownlint)
+
+In addition, there are numerous projects that hack around other languages to make them work in ESLint, such as:
+
+* [GraphQL-ESLint](https://github.com/B2o5T/graphql-eslint)
+* [eslint-plugin-json](https://www.npmjs.com/package/eslint-plugin-json)
+* [typescript-eslint](https://typescript-eslint.io)
+
+Instead of forcing others to rewrite what is core linting functionality, or asking people to produce an ESTree format AST for some language other than JavaScript, we can make everyone's lives easier by providing formal language support such that languages can be plugged in to ESLint easily.
+
+## Detailed Design
+
+This proposal consists of the following changes:
+
+1. Allows plugins to publish languages in addition to rules, processors, etc.
+1. Allows end users to select which language to use in their config files.
+1. Defines an API for defining languages.
+1. Changes the ESLint core to consume language definitions instead of assuming JavaScript is always used.
+
+**Prerequisites:** This proposal requires the flat config system and cannot be used with the eslintrc config system.
+
+**Note:** This design is intended to work both in the current version of ESLint and also in the "rewrite" version, which I'll refer to as ESLint X to distinguish between the two. Not all features of this proposal will be available in both versions. 
+
+### Languages in plugins and configs
+
+Languages can be defined in plugins by using the `languages` key and giving the language a name. This allows multiple languages to be contained in a single plugin if desired. For example:
+
+```js
+// inside a plugin called eslint-plugin-lang
+exports default {
+
+    languages: {
+        lang1: {
+            // language object
+        },
+        lang2: {
+            // language object
+        }
+    },
+    processors: {
+        // ..
+    },
+    rules: {
+        // ...
+    }
+};
+```
+
+Here, `lang1` and `lang2` are languages defined by the plugin. End users can specify which language to use via the `language` key in a config object, such as:
+
+
+```js
+// inside eslint.config.js
+import lang from "eslint-plugin-lang";
```

Comment:
I'd like to propose that the language use a new package name, like `eslint-lang-name`